### PR TITLE
ref(perf): Harden `useSpanMetricsSeries` data hook

### PR DESCRIPTION
**Pull request by gggritso**
`useSpanMetricsSeries` now matches `useSpanMetrics`:

- strict typing! Only genuine metrics fields are allowed in `yAxis`
- generic return type now takes `yAxis` into account
- same logic under-the-hood

**e.g.,**

Here's TypeScript's auto-complete being helpful:

Screenshot 2024-01-15 at 5 53 23 PM

## Changes

- Remove pointless comment
- `useLocation` --> `usePageFilters`
- Add strict `Aggregate` typing
- Add sum an an allow aggregate function
- Use interpolated types to declare all aggregates
- Add strict typing to `useSpanMetricsSeries`

### CODE REVIEW

File: static/app/views/performance/database/useSelectedDurationAggregate.tsx [hunk 1]
```diff
@@ -34,12 +33,16 @@ export function useSelectedDurationAggregate(): Result {
...
+  let selectedAggregate: Aggregate;
+  const aggregateFromURL = decodeScalar(
     location.query.aggregate,
     previouslySelectedAggregate
   );
 
-  if (!availableAggregates.includes(selectedAggregate)) {
+  // `availableAggregates` is typed `as const` so I have to cast to unknown to allow comparison to string
+  if ((availableAggregates as unknown as string[]).includes(aggregateFromURL)) {
+    selectedAggregate = aggregateFromURL as Aggregate;
```
Comment (AbhiPrasad) [inline static/app/views/performance/database/useSelectedDurationAggregate.tsx:44]: m: It was a bit hard for me to understand what the `includes` call was doing at first. What do you think of something like this?

```ts
const aggregateFromURL = decodeScalar(
  location.query.aggregate,
  previouslySelectedAggregate
);

const selectedAggregate = isAggregate(availableAggregates, aggregateFromURL)
  ? aggregateFromURL
  : DEFAULT_DURATION_AGGREGATE;

return [selectedAggregate, setSelectedAggregate];
```

where `isAggregate` is a helper like:

```ts
function isAggregate(
  availableAggregates: Aggregate[],
  maybeAggregate: string
): maybeAggregate is Aggregate {
  // couldn't think of a way around the cast here unfortunately, but still should be type safe!
  return availableAggregates.includes(maybeAggregate as Aggregate);
}
```

**Review by AbhiPrasad — APPROVED** [review]

File: static/app/views/performance/database/useAvailableDurationAggregates.tsx [hunk 2]
```diff
@@ -1,19 +1,18 @@
...
   const organization = useOrganization();
 
-  const availableAggregates: string[] = [];
+  const availableAggregates: Aggregate[] = [];
 
   availableAggregates.push(DEFAULT_DURATION_AGGREGATE);
 
   if (organization.features?.includes('performance-database-view-percentiles')) {
-    availableAggregates.push(...['p50', 'p75', 'p95', 'p99']);
+    availableAggregates.push(...(['p50', 'p75', 'p95', 'p99'] as Aggregate[]));
```
[the remark this answers is not in the document]
Comment (gggritso) [reply static/app/views/performance/database/useAvailableDurationAggregates.tsx:15]: No dice, `availableAggregates` is typed as `Aggregate[]` and `['p50', 'p75', 'p95', 'p99']` is widened to `string[]` so I need to cast manually here. I can remove the optional chaining though

My favourite alternative is actually

```ts
    availableAggregates.push(
      ...['p50' as const, 'p75' as const, 'p95' as const, 'p99' as const]
    );
```

which looks a little wacky, but does the least lying about types

File: static/app/views/performance/database/useSelectedDurationAggregate.tsx (hunk 1 above)
Comment (gggritso) [reply static/app/views/performance/database/useSelectedDurationAggregate.tsx:44]: Works for me! I had a version like that at some point but it got lost in between refactors. I'll bring it back 👍🏻

File: static/app/views/performance/database/useAvailableDurationAggregates.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (AbhiPrasad) [reply static/app/views/performance/database/useAvailableDurationAggregates.tsx]: when I used varadic args with `.push` TS didn't complain for me. See TS playground: https://www.typescriptlang.org/play?#code/MYewdgzgLgBAwgeQKoDkAqBRASgfQIIDiBWGBemAyjALwwDaA5BAK4C2DANDAwIYBuAc07dWASzDCGrHgA9JABwCMABmUMAujB4QYoSFADcAWABQe6DAAiASQpos1gEJI01hCnxESZSjXoN5AFY1LgCAdkCFAE5I0PkoqI0tHXNDU1NUmEJiUnIMKlo6ADoSxFRMXGzvPIouEqKbOwdnV3dPHJ98zW1dcGhjE1MoAE95AFMsgQEAJzGBHigJ2gAKEfGQADMsr1zKAEo6MDYAIzHp9QGh0YmsMZYAG1haPCnZ+cW6C-SzPthLDAAYngkAAZNA4SxILDkNweKq7DAALkmMzmCyW3H4QkuJg2zDAwCgonAMGYEDGeD4PFE9x4x3uY0szGmC2JYBeqPed2We2RtwesAA3qYAPQimASzIgabzMCiABerJJtDJYwQMp4csVRPAPJxkt+WipNLpDI5b3REGR5rRH00hS+gxMEv41Np9Ipr1tdyK8jJAAtlv8gaDwZDoa04TtOnscWKJTBRFtltLZQqlWAihsxgtmXcAPxFcTAe7MAAm3ICZw20ukBLGAFoyws6dpG3xRGMAO4N8bTYBjMBEhkQBh7PYwYXOhOuk0em1ciC+gPLALBBQRaKxbjxRKx0XiiUAX2+EtmUGZYCNbtNns5loGJ6d5hADKK9xAAmWqspN49TJZHV2S9RceT2IA

[the remark this answers is not in the document]
Comment (gggritso) [reply static/app/views/performance/database/useAvailableDurationAggregates.tsx]: Oh, I see what you mean, yes! No widening in that case 👍🏻 fixed

**Review by AbhiPrasad — APPROVED** [review]

**Pull request merged**